synapse/handlers/room.py

```diff
@@ -1167,12 +1167,13 @@ async def filter_evts(events: List[EventBase]) -> List[EventBase]:
         # XXX: why do we return the state as of the last event rather than the
         # first? Shouldn't we be consistent with /sync?
         # https://github.com/matrix-org/matrix-doc/issues/687
+        event_id_to_get_state_from = last_event_id
 
         state = await self.state_store.get_state_for_events(
-            [last_event_id], state_filter=state_filter
```

Comment:
For `/context`, it only gives you the state from the last event in `events_after` instead of the `event_id` itself. This means that often times, you end up missing the historical floating state for the historical message and instead get some of the real resolved state which doesn't have any of the member events we want.

I was messing around with changing this to return the state for the `event_id` specified but using the last event is part of the spec.

I could change the which events are fetched in `events_before`/`events_after` when specifying a historical event so it only returns itself 🤷. Then it would be the last event of itself.